### PR TITLE
feature-and-test: add has_followed on friendships page with tests

### PR DESCRIPTION
**Pull request by W-Shih**
- feature: add has_followed on friendships page
- test: add tests for showing has_followed on the friendships page
- TODO: this feature causes N + 1 query problem, need to improve this by cache

### CODE REVIEW

File: friendships/api/tests.py [hunk 1]
```diff
@@ -293,3 +294,106 @@ def test_friendship_pagination(self):
...
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data['results']), page_size)
+        self.assertEqual(response.data['count'], num_followers)
+        for i in range(page_size):
+            user_id = response.data['results'][i]['from_user']['id']
+            has_kd_followed = response.data['results'][i]['has_followed']
+            if user_id % 2 == 0:
+                self.assertEqual(has_kd_followed, True)
+            else:
+                self.assertEqual(has_kd_followed, False)
```
Comment (wen-wayne-shih) [inline friendships/api/tests.py:319]:
```suggestion
        for result in response.data['results']:
            user_id = result['from_user']['id']
            has_kd_followed = (user_id % 2 == 0)
            self.assertEqual(result['has_followed'], has_kd_followed)
```

File: friendships/api/tests.py [hunk 2]
```diff
@@ -293,3 +294,106 @@ def test_friendship_pagination(self):
...
+        self.assertEqual(response.data['page_number'], 1)
+
+        # anonymous hasn't followed any user  <Wayne Shih> 03-Apr-2022
+        response = self.anonymous_client.get(url, {'page': 2})
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.data['count'], num_followers)
+        self.assertEqual(len(response.data['results']), page_size)
+        for i in range(page_size):
+            has_anonymous_followed = response.data['results'][i]['has_followed']
+            self.assertEqual(has_anonymous_followed, False)
```
Comment (wen-wayne-shih) [inline friendships/api/tests.py:334]:
```suggestion
        for result in response.data['results']:
            self.assertEqual(result['has_followed'], False)
```

File: friendships/api/tests.py [hunk 3]
```diff
@@ -293,3 +294,106 @@ def test_friendship_pagination(self):
...
+                Friendship.objects.create(from_user=self.kb24, to_user=user)
+
+        # kd has followed all his followings  <Wayne Shih> 03-Apr-2022
+        response = self.kd35_client.get(url)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data['results']), page_size)
+        self.assertEqual(response.data['count'], num_followings)
+        for i in range(page_size):
+            has_kd_followed = response.data['results'][i]['has_followed']
+            self.assertEqual(has_kd_followed, True)
```
Comment (wen-wayne-shih) [inline friendships/api/tests.py:359]: same comment as above

File: friendships/api/tests.py [hunk 4]
```diff
@@ -293,3 +294,106 @@ def test_friendship_pagination(self):
...
+        self.assertEqual(response.data['has_next'], True)
+        self.assertEqual(response.data['previous'], None)
+        self.assertEqual('?page=2' in response.data['next'], True)
+        for i in range(page_size):
+            user_id = response.data['results'][i]['user']['id']
+            has_kb_followed = response.data['results'][i]['has_followed']
+            if user_id % 2 == 0:
+                self.assertEqual(has_kb_followed, True)
+            else:
+                self.assertEqual(has_kb_followed, False)
```
Comment (wen-wayne-shih) [inline friendships/api/tests.py:384]: same comment as above

File: friendships/api/tests.py [hunk 5]
```diff
@@ -293,3 +294,106 @@ def test_friendship_pagination(self):
...
+
+        # anonymous hasn't followed any user  <Wayne Shih> 03-Apr-2022
+        response = self.anonymous_client.get(url, {'page': 2})
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        for i in range(page_size):
+            has_anonymous_followed = response.data['results'][i]['has_followed']
+            self.assertEqual(has_anonymous_followed, False)
```
Comment (wen-wayne-shih) [inline friendships/api/tests.py:391]: same comment as above

**Review by wen-wayne-shih — CHANGES_REQUESTED** [review]

File: friendships/api/tests.py (hunk 1 above)
Comment (W-Shih) [reply friendships/api/tests.py]: Resolved as per suggestion

File: friendships/api/tests.py (hunk 2 above)
Comment (W-Shih) [reply friendships/api/tests.py]: Resolved as per suggestion

File: friendships/api/tests.py (hunk 3 above)
Comment (W-Shih) [reply friendships/api/tests.py]: Resolved as per suggestion

File: friendships/api/tests.py (hunk 4 above)
Comment (W-Shih) [reply friendships/api/tests.py]: Resolved as per suggestion

File: friendships/api/tests.py (hunk 5 above)
Comment (W-Shih) [reply friendships/api/tests.py]: Resolved as per suggestion

**Review by wen-wayne-shih — APPROVED** [review]
LGTM

**Pull request merged**